jax/experimental/sparse/ops.py

```diff
@@ -1175,6 +1175,139 @@ def impl(A, B, indices):
+      in_axes = 0
+    else:
+      raise NotImplementedError("unequal batches in batched dims")
+    func = vmap(func, in_axes=in_axes)
+  return func(lhs_data, lhs_indices, rhs_data, rhs_indices)
+
+@bcoo_spdot_general_p.def_abstract_eval
+def _bcoo_spdot_general_abstract_eval(lhs_data, lhs_indices, rhs_data, rhs_indices, *, lhs_shape, rhs_shape, dimension_numbers):
+  lhs_n_batch, lhs_n_sparse, lhs_n_dense = _validate_bcoo(lhs_data, lhs_indices, lhs_shape)
+  rhs_n_batch, rhs_n_sparse, rhs_n_dense = _validate_bcoo(rhs_data, rhs_indices, rhs_shape)
+  (lhs_contracting, rhs_contracting), (lhs_batch, rhs_batch) = dimension_numbers
+
+  # Check for proper dimension_numbers
+  for dims in [lhs_contracting, rhs_contracting, lhs_batch, rhs_batch]:
+    assert len(dims) == len(set(dims))
+  assert not set(lhs_contracting).intersection(lhs_batch)
+  assert not set(rhs_contracting).intersection(rhs_batch)
+  assert [lhs_shape[d] for d in lhs_contracting] == [rhs_shape[d] for d in rhs_contracting]
+  assert [lhs_shape[d] for d in lhs_batch] == [rhs_shape[d] for d in rhs_batch]
+
+  if not (lhs_n_dense == rhs_n_dense == 0):
+    # TODO(jakevdp): handle dense dimensions
+    raise NotImplementedError("bcoo_spdot_general with dense dimensions.")
+
+  if not (lhs_n_sparse == rhs_n_sparse == 1):
+    raise NotImplementedError("bcoo_spdot_general with more than one sparse dimension.")
+
+  if any(dim >= lhs_n_batch for dim in lhs_batch) or any(dim > rhs_n_batch for dim in rhs_batch):
+    raise NotImplementedError("bcoo_spdot_general: batch_dims must correspond to batch dimensions of the sparse representation.")
+
+  if tuple(lhs_contracting) != (lhs_n_batch,) or tuple(rhs_contracting) != (rhs_n_batch,):
+    raise NotImplementedError("bcoo_spdot_general only supports contraction of sparse indices.")
+
+  out_shape = (
+    tuple(lhs_shape[dim] for dim in lhs_batch) +
+    tuple(lhs_shape[i] for i in range(lhs_n_batch) if i not in lhs_batch) +
+    tuple(rhs_shape[i] for i in range(rhs_n_batch) if i not in rhs_batch))
+  out_dtype = jnp.promote_types(lhs_data.dtype, rhs_data.dtype)
+
+  return core.ShapedArray(out_shape, dtype=out_dtype)
+
+def _bcoo_spdot_general_batch_rule(batched_args, batch_dims, *, dimension_numbers, lhs_shape, rhs_shape):
+  lhs_data, lhs_indices, rhs_data, rhs_indices = batched_args
+  batch_dims = list(batch_dims)
+  batch_size = max(0 if dim is None else arg.shape[dim]
+                   for arg, dim in zip(batched_args, batch_dims))
+  if batch_dims[0] is None:
+    lhs_data = lhs_data[None]
+    batch_dims[0] = 0
+  if batch_dims[1] is None:
+    lhs_indices = lhs_indices[None]
+    batch_dims[1] = 0
+  assert batch_dims[0] == batch_dims[1] == 0
+  if batch_dims[2] is None:
+    rhs_data = rhs_data[None]
+    batch_dims[2] = 0
+  if batch_dims[3] is None:
+    rhs_indices = rhs_indices[None]
+    batch_dims[3] = 0
+  # TODO: handle batchings along different axes?
+  assert all(dim == 0 for dim in batch_dims)
```

Comment:
Maybe a `raise NotImplementedError` instead? I always suggest this kind of thing because in theory someone could run with `python -O` and elide all assertions.

Reply:
Oh, good call. Done!